src/decoder.ts

```diff
@@ -0,0 +1,595 @@
+      `Platform '${vendor}:${arch}' is not installed`,
+      { sketchPath, fqbn }
+    );
+  }
+  if (!supportedArchitectures.has(arch)) {
+    throw new DecodeParamsError(`Unsupported board: '${fqbn}'`, {
+      sketchPath,
+      fqbn,
+    });
+  }
+  if (!compileSummary) {
+    throw new DecodeParamsError(
+      'The summary of the previous compilation is unavailable. Compile the sketch',
+      {
+        sketchPath,
+        fqbn,
+      }
+    );
+  }
+  const { buildPath } = compileSummary;
+  const sketchFolderName = path.basename(sketchPath);
+  const [toolPath, elfPath] = await Promise.all([
+    findToolPath(boardDetails),
+    findElfPath(sketchFolderName, buildPath),
+  ]);
+  if (!elfPath) {
+    throw new DecodeParamsError(
+      `Could not detect the '.elf' file in the build folder`,
+      { sketchPath, fqbn }
+    );
+  }
+  if (!toolPath) {
+    throw new DecodeParamsError('Could not detect the DGB tool path', {
+      sketchPath,
+      fqbn,
+    });
+  }
+  return {
+    toolPath,
+    elfPath,
+    fqbn,
+    sketchPath,
+  };
+}
+
+export class DecodeParamsError extends Error {
+  constructor(
+    message: string,
+    private readonly partial: Pick<DecodeParams, 'fqbn' | 'sketchPath'>
+  ) {
+    super(message);
+    Object.setPrototypeOf(this, DecodeParamsError.prototype);
+  }
+
+  get fqbn(): string {
+    return this.partial.fqbn;
+  }
+
+  get sketchPath(): string {
+    return this.partial.sketchPath;
+  }
+}
+
+export interface DecodeOptions {
+  readonly signal?: AbortSignal;
+  readonly debug?: Debug;
+}
+export const defaultDecodeOptions = {
+  signal: neverSignal,
+  debug: decoderDebug,
+} as const;
+
+export interface GDBLine {
+  readonly address: string;
+  readonly line: string;
+}
+export function isGDBLine(arg: unknown): arg is GDBLine {
+  return (
+    typeof arg === 'object' &&
+    (<GDBLine>arg).address !== undefined &&
+    typeof (<GDBLine>arg).address === 'string' &&
+    (<GDBLine>arg).line !== undefined &&
+    typeof (<GDBLine>arg).line === 'string'
+  );
+}
+
+export interface ParsedGDBLine extends GDBLine {
+  readonly file: string;
+  readonly method: string;
+}
+export function isParsedGDBLine(gdbLine: GDBLine): gdbLine is ParsedGDBLine {
+  return (
+    (<ParsedGDBLine>gdbLine).file !== undefined &&
+    typeof (<ParsedGDBLine>gdbLine).file === 'string' &&
+    (<ParsedGDBLine>gdbLine).method !== undefined &&
+    typeof (<ParsedGDBLine>gdbLine).method === 'string'
+  );
+}
+
+/**
+ * Register address in the `0x` format. For example, `'0x4020104e'`.
+ */
+export type Address = string;
+/**
+ * Successfully decoded register address, or the address.
+ */
+export type Location = GDBLine | Address;
+
+export interface DecodeResult {
+  readonly exception: [message: string, code: number] | undefined;
+  readonly registerLocations: Record<string, Location>;
+  readonly stacktraceLines: GDBLine[];
+  readonly allocLocation: [line: Location, size: number] | undefined;
```

Comment:
Rename `line` to `location`.

Reply:
https://github.com/dankeboy36/esp-exception-decoder/pull/1/commits/e05fb264090762876206d3e96f08d0e00786493a